src/lib/core/CHIPTLV.h

```diff
@@ -777,6 +777,13 @@ class DLL_EXPORT TLVReader
      */
     void * AppData;
 
+    /**
+     * Set TLV BackingStore
+     *
+     * @param[in] aBackingStore             The TLVBackingStore in TLV Reader
+     */
+    void SetBackingStore(TLVBackingStore * aBackingStore) { mBackingStore = aBackingStore; }
```

Comment:
This doesn't affect the `mMaxLen` of the reader?  Why not?

Looking at #4800 we want this because there we have a linked list of backing stores and want to move on to the next one.  But why can't we just have a backing store that contains a linked list of things and move on to the next one as needed?  Then we'd just have a backing store that implements GetNextBuffer as it sees fit, and we wouldn't need this back-door api, it seems like.

Reply:
Hi, @bzbarsky-apple 
The key code is here
https://github.com/project-chip/connectedhomeip/blob/master/src/lib/core/CHIPTLVReader.cpp#L778

When previous tlv reader reaches the end, then it look for next buffer via mBackingStore->GetNextBuffer(*this, mReadPoint, bufLen), then it update mBufEnd = mReadPoint + bufLen;

Reply:
https://github.com/project-chip/connectedhomeip/blob/master/src/lib/core/CHIPTLVReader.cpp#L784

// Cap mBufEnd so that we don't read beyond the user's specified maximum length, even
// if the underlying buffer is larger.

Previously tlv is designed to use this way to link to new buffer and cap the maximum size

Reply:
another way to remove this setBackingStore is to add additional backstore parameter for GetNextBuffer https://github.com/project-chip/connectedhomeip/blob/1f7e79bd0f5554342d418fe61163eaec18a67993/src/lib/core/CHIPTLV.h#L2246

I would imagine the case, sometimes the next storage would not be the one in linked list, and could be any one user specify and offer.

Reply:
This last sounds like what I was expecting, yes.

It sounds like in practice `maxLen` is just `UINT32_MAX` in our case which is why it doesn't matter too much what we do with the backing store?  But conceptually, the backing store is just something; the fact that it has three separate circular buffers under it is an implementation detail that should imo be handled entirely in `GetNextBuffer`.

Reply:
just update the pr, the reference IM event logic is 
https://github.com/yunhanw-google/connectedhomeip/blob/feature/add_im_event/src/app/reporting/LoggingManagement.h#L107

I add three private virtual methods in CircularEventBuffer.
    void SetNextBackingStore(chip::TLV::TLVBackingStore * apBackingStore) override { mpNextBackingStore = apBackingStore; }
    chip::TLV::TLVBackingStore * GetNextBackingStore() override { return mpNextBackingStore; }
    void ClearNextBackingStore() override { mpNextBackingStore = nullptr; };

TLVReader call ensure data, then call GetNextBuffer, CircularEventBuffer::GetNextBuffer in CircularEventBuffer check whether current buffer is consumed and whether to move to next circular buffer(prev pointer saved in circular event buffer), then if it need move, it will call SetNextBackingStore, tlvreader can use GetNextBackingStore to obtain next backing store in tlvreader
https://github.com/yunhanw-google/connectedhomeip/blob/feature/add_im_event/src/app/reporting/LoggingManagement.cpp#L954